cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -165,71 +160,110 @@ func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+			return mig.Config, nil
 		}
 	}
 	// Instance doesn't belong to any configured mig.
 	return nil, nil
 }
 
-// RegenerateInstancesCache triggers instances cache regeneration under lock.
-func (gc *GceCache) RegenerateInstancesCache() error {
+// RegenerateInstancesCache triggers instances cache regeneration for single MIG under lock.
+func (gc *GceCache) RegenerateInstanceCacheForMig(migRef GceRef) error {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
-
-	return gc.regenerateCache()
+	return gc.regenerateInstanceCacheForMigNoLock(migRef)
 }
 
-// internal method - should only be called after locking on cacheMutex.
-func (gc *GceCache) regenerateCache() error {
-	newInstancesCache := make(map[GceRef]Mig)
+func (gc *GceCache) regenerateInstanceCacheForMigNoLock(migRef GceRef) error {
+	klog.V(4).Infof("Regenerating MIG information for %s", migRef.String())
 
-	for _, migInfo := range gc.GetMigs() {
-		mig := migInfo.Config
-		klog.V(4).Infof("Regenerating MIG information for %s", mig.GceRef().String())
+	// Here we are on purpose not deleting old entries pointing to migRef from gc.instanceRefToMig.
+	// They are no harm as same GceRef should not be used for new instance and instance should not move from one Mig to
+	// the other.
+	// And garbage-collection will be enforces as whole cache is invalidated hourly.
 
-		basename, err := gc.GceService.FetchMigBasename(mig.GceRef())
+	basename, err := gc.GceService.FetchMigBasename(migRef)
+	if err != nil {
+		return err
+	}
+	gc.updateMigBasename(migRef, basename)
+
+	instances, err := gc.GceService.FetchMigInstances(migRef)
+	if err != nil {
+		klog.V(4).Infof("Failed MIG info request for %s: %v", migRef.String(), err)
+		return err
+	}
+	for _, instance := range instances {
+		instanceRef, err := GceRefFromProviderId(instance.Id)
 		if err != nil {
 			return err
 		}
-		gc.updateMigBasename(mig.GceRef(), basename)
+		gc.instanceRefToMigRef[instanceRef] = migRef
+	}
+	return nil
+}
+
+// RegenerateInstancesCache triggers instances cache regeneration under lock.
+func (gc *GceCache) RegenerateInstancesCache() error {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.regenerateInstancesCacheNoLock()
+}
 
-		instances, err := gc.GceService.FetchMigInstances(mig.GceRef())
+func (gc *GceCache) regenerateInstancesCacheNoLock() error {
+	gc.invalidateInstanceCacheNoLock()
+	for migRef := range gc.migs {
+		err := gc.regenerateInstanceCacheForMigNoLock(migRef)
 		if err != nil {
-			klog.V(4).Infof("Failed MIG info request for %s: %v", mig.GceRef().String(), err)
 			return err
 		}
-		for _, instance := range instances {
-			gceRef, err := GceRefFromProviderId(instance.Id)
-			if err != nil {
-				return err
-			}
-			newInstancesCache[*gceRef] = mig
-		}
 	}
-
-	gc.instancesCache = newInstancesCache
 	return nil
 }
 
+// InvalidateInstanceCache invalidates whole InstanceRef->MigRef cache
+func (gc *GceCache) InvalidateInstanceCache() {
```

Comment:
Do we use it anywhere?

Reply:
Only the `NoLock` version. I will remove the public one.